classes/SchemaValidator.cls

```diff
@@ -0,0 +1,82 @@
+/*
+ * Validates SObject types, fields, and relationships
+ * @author github@cloudyjohnny
+ */
+public class SchemaValidator {
+    private Schema.SObjectType objectType;
+
+    public SchemaValidator(Schema.SObjectType objectType) {
+        this.objectType = objectType;
+    }
+
+    public void validateField(SObjectField field) {
+        this.validateField(field.getDescribe().getName());
+    }
+
+    public void validateField(String fieldName) {
+        if (!this.objectType.getDescribe().fields.getMap().containsKey(fieldName)) {
+            throw new InvalidFieldException (
+                String.format(
+                    InvalidFieldException.MESSAGE_TEMPLATE, 
+                    new List<String> { fieldName, this.objectType.getDescribe().getName() }
+                )
+            );
+        }
+    }
+
+    public void validateSObjectType(SObject record) {
+        if(record.getSObjectType() != this.objectType) {
+            throw new InvalidSObjectTypeException(
+                String.format(
+                    InvalidSObjectTypeException.MESSAGE_TEMPLATE,
+                    new List<String> {
+                        this.objectType.getDescribe().getName(),
+                        record.getSObjectType().getDescribe().getName()
+                    }
+                )
+            );
+        }
+    }
+
+    public void validateChildRelationship(String relationshipName, Schema.SObjectType childType) {
+        List<Schema.ChildRelationship> childRelationships = this.objectType.getDescribe().getChildRelationships();
+        for(Schema.ChildRelationship currRelationship : childRelationships) {
+            if (currRelationship.getRelationshipName() == relationshipName && currRelationship.getChildSObject() == childType) {
+                return;
+            }
+        }
```

Comment:
In case there are multiple child relationships in a query, it may be better to create a map with a key type of `Schema.ChildRelationship` and a value type of 'Schema.SObjectType'. The main repository will only need to call `this.objectType.getDescribe().getChildRelationships();` once and for the remaining validations, you could just check if the child relationship given matches to a non-null value. Thoughts?

Reply:
yeah I think that's a good idea. I think should have memoization of all of these describe calls since they aren't performant